sql/catalyst/src/main/scala/org/apache/spark/sql/connector/catalog/CatalogManager.scala

```diff
@@ -85,33 +91,53 @@ class CatalogManager(conf: SQLConf, defaultSessionCatalog: TableCatalog) extends
 
   def currentNamespace: Array[String] = synchronized {
     _currentNamespace.getOrElse {
-      currentCatalog.map { catalogName =>
-        getDefaultNamespace(catalog(catalogName))
-      }.getOrElse(Array("default")) // The builtin catalog use "default" as the default database.
+      // For session catalog, the current namespace is kept in `SessionCatalog`. There are many
+      // commands that do not support v2 catalog API. They ignore the current catalog and blindly
+      // go to `SessionCatalog`. This means, we must keep track of the current namespace of session
+      // catalog even if the current catalog is not session catalog. `CatalogManager` only tracks
+      // the current namespace of the current catalog.
+      if (currentCatalog.name() == SESSION_CATALOG_NAME) {
+        Array(v1SessionCatalog.getCurrentDatabase)
+      } else {
+        getDefaultNamespace(currentCatalog)
+      }
     }
   }
 
   def setCurrentNamespace(namespace: Array[String]): Unit = synchronized {
-    _currentNamespace = Some(namespace)
+    // For session catalog, the current namespace is kept in `SessionCatalog`. There are many
+    // commands that do not support v2 catalog API. They ignore the current catalog and blindly
+    // go to `SessionCatalog`. This means, we must keep track of the current namespace of session
+    // catalog even if the current catalog is not session catalog. `CatalogManager` only tracks
+    // the current namespace of the current catalog.
+    if (currentCatalog.name() == SESSION_CATALOG_NAME) {
+      if (namespace.length != 1) {
+        throw new NoSuchNamespaceException(namespace)
+      }
+      v1SessionCatalog.setCurrentDatabase(namespace.head)
+    } else {
+      _currentNamespace = Some(namespace)
+    }
   }
 
-  private var _currentCatalog: Option[String] = None
+  private var _currentCatalogName: Option[String] = None
 
-  // Returns the name of current catalog. None means the current catalog is the builtin catalog.
-  def currentCatalog: Option[String] = synchronized {
-    _currentCatalog.orElse(conf.defaultV2Catalog)
+  def currentCatalog: CatalogPlugin = synchronized {
+    _currentCatalogName.map(catalogName => catalog(catalogName))
+      .orElse(defaultCatalog)
+      .getOrElse(v2SessionCatalog)
   }
 
   def setCurrentCatalog(catalogName: String): Unit = synchronized {
-    _currentCatalog = Some(catalogName)
+    _currentCatalogName = Some(catalogName)
     _currentNamespace = None
```

Comment:
Don't we need to reset the namespace (set to default database) in `v1SessionCatalog` to make the behavior consistent when catalog is changed?

Reply:
This is a behavior we can discuss: a user sets the current database of the session catalog, then switch to a custom catalog, shall we reset the current database of the session catalog? Think about the following commands:
```
USE myDB
SET CURRENT CATALOG myCata
ANALYZE TABLE t
```
shall we analyze `myDB.t` or `default.t` or fail?

also cc @gatorsmile @dongjoon-hyun @viirya

Reply:
I think the current namespace should be either 1) reset to default or 2) maintained per catalog, but not carried over when a catalog is switched.

For example,
1) reset to default:
```SQL 
# Starts with session catalog
SELECT current_database() # default
USE db
SELECT current_database() # db

# switch catalog
USE CATALOG myCat
SELECT current_database() # [] (default)
USE ns1.ns2
SELECT current_database() # ns1.ns2

# switch back to different catalogs. namespace is reset.
USE CATALOG session
SELECT current_database() # default
USE CATALOG myCat
SELECT current_database() # [] (default)
```

2) previous namespace is maintained per catalog:
```SQL 
# Starts with session catalog
SELECT current_database() # default
USE db
SELECT current_database() # db

# switch catalog
USE CATALOG myCat
SELECT current_database() # [] (default)
USE ns1.ns2
SELECT current_database() # ns1.ns2

# switch back to different catalogs. namespace is preserved.
USE CATALOG session
SELECT current_database() # db
USE CATALOG myCat
SELECT current_database() # ns1.ns2
```

Reply:
+1 for resetting it when we switch.

Reply:
resetting it when switching is definitely reasonable, but what I was talking about is v1 command:
```
# Starts with session catalog
SELECT current_database() # default
ANALYZE TABLE t # analyze default.t from session catalog

USE db
SELECT current_database() # db
ANALYZE TABLE t # analyze db.t from session catalog

# switch catalog
USE CATALOG myCat
SELECT current_database() # [] (default namespace of myCat)
ANALYZE TABLE t
```
For the last `ANALYZE TABLE`,it's unclear what the behavior should be. There are 3 options:
1. We should look up the table from `myCat`, so this should either throw NoSuchTable exception, or throw unsupported exception because we don't have an analyze table API in DS v2.
2. V1 commands should still go with the v1 code path. We should analyze table `default.t` because the current catalog is not session catalog and we don't know what the currennt database of session catalog is now.
3. V1 commands should still go with the v1 code path. We should analyze table `db.t` because this is the most recent current database of  the session catalog.

Reply:
Sorry, @cloud-fan . I missed your point. For that last `ANALYZE TABLE`, I thought `USE CATALOG` should follow Option 1.

Reply:
Personally I also think option 1 is better, but it's a lot of work. We need to create statement logical plans (e.g. `DeleteFromStatement`) and related resolution rules for all commands that needs to resolve tables even if they don't support v2 APIs. An initial list: ANALYZE TABLE, CACHE TABLE, REFRESH TABLE, ALTER TABLE ADD PARTITION, ALTER TABLE SET SERDE, REPAIR TABLE, ...

My [refactoring](https://github.com/apache/spark/pull/25747) should make this easier, but it's still a lot of work to be done here.

Reply:
+1 for Option 1. I think we can make the behavior consistent when we switch catalogs (reset to default). But if Option 1 is a lot of work, we can go with Option 2 (`default.t`) until we create rules for all commands that don't support v2 APIs (I can help with this).

Reply:
cool! I'll go with option 2 here, with a TODO saying that we want to do option 1 in the future.

I'll create JIRA tickets for the followup work so that we can have people working on it in parallel,

Reply:
Got it, @cloud-fan .

Reply:
+1 Agree on that option 1 is more reasonable.